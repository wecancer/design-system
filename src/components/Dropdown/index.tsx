import React, {useState, useRef} from 'react'
import styled, {css} from 'styled-components'

import OutsideEvent from '../OutsideEvent'

type AxisX = 'left' | 'right'

const Wrapper = styled(OutsideEvent)`
  position: relative;
  display: inline-block;
`

const TriggerContainer = styled.div``

const BoxContainer = styled.div<{triggerHeight: number; axisX: AxisX}>`
  ${({triggerHeight, axisX}) => css`
    position: absolute;
    top: ${triggerHeight + 10}px;
    ${axisX === 'left'
      ? css`
          left: 0;
        `
      : css`
          right: 0;
        `}
  `}
`

export type ActionParams = {
  isOpen: boolean
  handleClose(): void
  handleToggle(): void
}

export type Props = {
  axisX?: AxisX
  trigger(params: ActionParams): React.ReactNode
  children(params: ActionParams): React.ReactNode
  className?: string
}

const Dropdown = ({axisX = 'left', children, trigger, ...props}: Props) => {
  const refTrigger = useRef<HTMLDivElement>(null)
  const [isOpen, setOpen] = useState(false)
  const handleClose = () => setOpen(false)
  const handleToggle = () => setOpen(!isOpen)

  return (
    <Wrapper onClickOutside={handleClose} {...props}>
      <TriggerContainer ref={refTrigger}>
        {trigger({isOpen, handleClose, handleToggle})}
      </TriggerContainer>
      {isOpen && (
        <BoxContainer
          axisX={axisX}
          triggerHeight={refTrigger.current?.offsetHeight || 0}
        >
          {children({isOpen, handleClose, handleToggle})}
        </BoxContainer>
      )}
    </Wrapper>
  )
}

export default Dropdown
