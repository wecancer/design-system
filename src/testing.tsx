import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
