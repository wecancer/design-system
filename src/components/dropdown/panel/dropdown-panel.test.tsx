import Dropdown from '.'
import { render, fireEvent } from '../../../testing'

describe('<DropdownPanel />', () => {
  const options = [
    {
      label: 'item 1',
      id: '01',
    },
    {
      label: 'item 2',
      id: '02',
    },
    {
      label: 'item 3',
      id: '03',
    },
  ]

  it('should render snapshot correctly', () => {
    const { container } = render(
      <>
        <Dropdown label="empty options" items={[]} />
        <Dropdown label="with " items={options} />
      </>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should open the options when the lavel is clicked', () => {
    const { getByText, getByRole } = render(
      <Dropdown label="my label" items={options} />,
    )
    const label = getByText('my label')

    fireEvent.click(label)
    expect(getByRole('listbox')).toBeInstanceOf(HTMLElement)
  })
})
