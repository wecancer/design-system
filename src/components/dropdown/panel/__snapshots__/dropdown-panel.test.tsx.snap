// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownPanel /> should render snapshot correctly 1`] = `
<div
  class="sc-bdvvtL iqkGDp"
>
  <div
    class="sc-gsDKAQ"
  >
    <div
      class="sc-iCfMLu iGWyQt"
      role="button"
      tabindex="0"
    >
      <p
        class="sc-furwcr dlyZeQ"
      >
        empty options
      </p>
    </div>
  </div>
</div>
`;
